pslab/external/hcsr04.py

```diff
@@ -0,0 +1,125 @@
+"""Ultrasonic distance sensors."""
+import time
+
+from pslab.instrument.logic_analyzer import LogicAnalyzer
+from pslab.instrument.waveform_generator import PWMGenerator
+from pslab.serial_handler import SerialHandler
+
+
+class HCSR04(LogicAnalyzer, PWMGenerator):
+    """Read data from ultrasonic distance sensor HC-SR04/HC-SR05.
+
+    These sensors can measure distances between 2 cm to 4 m (SR04) / 4.5 m
+    (SR05).
+
+    Sensors must have separate trigger and echo pins. First a 10 µs pulse is
+    output on the trigger pin. The trigger pin must be connected to the TRIG
+    pin on the sensor prior to use.
+
+    Upon receiving this pulse, the sensor emits a sequence of sound pulses, and
+    the logic level of its echo pin is also set high.  The logic level goes LOW
+    when the sound packet returns to the sensor, or when a timeout occurs.
+    Timeout occurs if no echo is received within the time slot determinded by
+    the sensor's maximum range.
```

Comment:
@CloudyPadmal Please check if this is correct. The datasheet is unclear and different online sources contradict each other on this point.

Reply:
Looks like the sensor I'm using (HC-SR04P) keeps the echo pin high without any internal timeout. 

When I tried the code snippet from [firmware-pr](https://github.com/fossasia/pslab-firmware/pull/118#issuecomment-1076290022), the LA capture list always miss the last entry (could be that it's missing the last falling edge).

Reply:
Does the ECHO pin eventually go back to LOW? IF so, does the method capture the final edge if you give it more time, say an extra 60 ms or so?

Reply:
I will check and confirm

Reply:
Hi Alex, so I checked the ECHO pin status using an oscilloscope. Looks like it eventually goes to LOW after 200 ms.

<table>
<tr>
<td><img src="https://user-images.githubusercontent.com/14261304/161245231-4c6961e7-f869-4830-910b-102cc39d2dc0.jpg" /></td>
<td><img src="https://user-images.githubusercontent.com/14261304/161245245-ddf8c7c8-d371-4ad3-9c45-2c3c105e6757.jpg" /></td>
</tr>
</table>

This causes it to throw the following error (when it is a timeout)

```
<script_path>: RuntimeWarning: Mean of empty slice.
  return speed_of_sound * high_times.mean() / 2 * 1e-6
python3.7/site-packages/numpy/core/_methods.py:189: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
nan
```